File: public/cpp/096-kth-largest-element.cpp
/*
 * Problem: Kth Largest Element in an Array (LeetCode 215)
 * Link: https://leetcode.com/problems/kth-largest-element-in-an-array/
 * Difficulty: Medium
 * Category: Sorting and Searching
 * 
 * Description:
 * Given an integer array nums and an integer k, return the kth largest element.
 * Note that it is the kth largest element in sorted order, not the kth distinct element.
 * 
 * Example 1:
 * Input: nums = [3,2,1,5,6,4], k = 2
 * Output: 5
 * 
 * Example 2:
 * Input: nums = [3,2,3,1,2,4,5,5,6], k = 4
 * Output: 4
 */

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

/*
 * APPROACH 1: QUICKSELECT (OPTIMAL)
 * Time: O(n) average, O(n²) worst, Space: O(1)
 */
class Solution1 {
private:
    int partition(vector<int>& nums, int left, int right) {
        int pivot = nums[right];
        int i = left;
        
        for (int j = left; j < right; j++) {
            if (nums[j] >= pivot) {
                swap(nums[i], nums[j]);
                i++;
            }
        }
        swap(nums[i], nums[right]);
        return i;
    }
    
    int quickSelect(vector<int>& nums, int left, int right, int k) {
        if (left == right) return nums[left];
        
        int pivotIndex = partition(nums, left, right);
        
        if (pivotIndex == k) return nums[pivotIndex];
        else if (pivotIndex < k) return quickSelect(nums, pivotIndex + 1, right, k);
        else return quickSelect(nums, left, pivotIndex - 1, k);
    }
    
public:
    int findKthLargest(vector<int>& nums, int k) {
        return quickSelect(nums, 0, nums.size() - 1, k - 1);
    }
};

/*
 * APPROACH 2: MIN HEAP
 * Time: O(n log k), Space: O(k)
 */
class Solution2 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> minHeap;
        
        for (int num : nums) {
            minHeap.push(num);
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }
        return minHeap.top();
    }
};

/*
 * APPROACH 3: SORTING
 * Time: O(n log n), Space: O(1)
 */
class Solution3 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end(), greater<int>());
        return nums[k - 1];
    }
};

/*
 * APPROACH 4: MAX HEAP
 * Time: O(n + k log n), Space: O(n)
 */
class Solution4 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int> maxHeap(nums.begin(), nums.end());
        
        for (int i = 0; i < k - 1; i++) {
            maxHeap.pop();
        }
        return maxHeap.top();
    }
};

/*
 * APPROACH 5: STL NTH_ELEMENT
 * Time: O(n) average, Space: O(1)
 */
class Solution5 {
public:
    int findKthLargest(vector<int>& nums, int k) {
        nth_element(nums.begin(), nums.begin() + k - 1, nums.end(), greater<int>());
        return nums[k - 1];
    }
};

void test(vector<int> nums, int k, int approach) {
    int result;
    
    switch(approach) {
        case 1: { Solution1 sol; result = sol.findKthLargest(nums, k); break; }
        case 2: { Solution2 sol; result = sol.findKthLargest(nums, k); break; }
        case 3: { Solution3 sol; result = sol.findKthLargest(nums, k); break; }
        case 4: { Solution4 sol; result = sol.findKthLargest(nums, k); break; }
        case 5: { Solution5 sol; result = sol.findKthLargest(nums, k); break; }
    }
    cout << result << "\n";
}

int main() {
    vector<int> nums = {3, 2, 1, 5, 6, 4};
    int k = 2;
    
    for (int i = 1; i <= 5; i++) {
        cout << "Approach " << i << ": ";
        test(nums, k, i);
    }
    return 0;
}

// Made with Bob

File: src/data/tutorials/096-kth-largest-element.js
/**
 * Tutorial content for #096 — Kth Largest Element (Quickselect focus)
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Find the kth largest element in an unsorted array. This is the same core problem as #084, but this version highlights the Quickselect algorithm as the primary solution. Note: "kth largest" means the kth in descending sorted order — so k=1 is the maximum, k=2 is the second largest, etc.`,
    example: `nums = [3,2,3,1,2,4,5,5,6], k = 4\nSorted descending: [6,5,5,4,3,3,2,2,1]\n4th largest = 4\n✅ Answer: 4`,
    keyInsight: `Quickselect: pick a pivot, partition the array so all elements ≥ pivot are on the left. If the pivot lands at index k-1 (0-based), it IS the kth largest. Otherwise recurse on the relevant half. Average O(n).`,
  },

  approaches: {
    'Quickselect': {
      intuition: `Choose a pivot element. Partition the array in-place: all elements LARGER than the pivot go left, smaller go right (for kth largest). If the pivot ends up at position k-1, we found the answer. If it's at position > k-1, the kth largest is in the left part; if < k-1, it's in the right part.\n\nWith a random pivot, this runs in O(n) on average.`,
      steps: [
        `Set left=0, right=n-1. Target index = k-1.`,
        `While left < right:`,
        `  Randomly pick pivot; swap to end.`,
        `  Partition: i=left. For j=left to right-1: if nums[j] >= pivot, swap nums[i] and nums[j], i++.`,
        `  Swap nums[i] and nums[right] (place pivot).`,
        `  If i == k-1: return nums[i].`,
        `  If i < k-1: left = i + 1.`,
        `  Else: right = i - 1.`,
        `Return nums[left].`,
      ],
      example: `nums=[3,2,3,1,2,4,5,5,6], k=4. Target index=3.\nPick pivot=4 (index 5).\nPartition (keep ≥4 on left): [5,5,6,4, | 2,3,1,2,3] pivot at index 3.\nPivot index==3==target → return nums[3]=4 ✅`,
      keyInsight: `O(n) average, O(n²) worst case, O(1) extra space (in-place). Use random pivot selection to avoid worst case. The gold-standard for "find kth element" when average speed matters.`,
    },

    'Min Heap of Size K': {
      intuition: `Maintain a min-heap of size k as you scan the array. Elements smaller than the heap's minimum get discarded; larger ones push out the minimum. After scanning, the heap's minimum is the kth largest.`,
      steps: [
        `Create empty min-heap.`,
        `For each num in nums:`,
        `  heap.add(num).`,
        `  If heap.size > k: heap.poll() (remove smallest).`,
        `Return heap.peek().`,
      ],
      example: `nums=[3,2,3,1,2,4,5,5,6], k=4\n\nAdd 3: [3]\nAdd 2: [2,3]\nAdd 3: [2,3,3]\nAdd 1: [1,2,3,3] size=4=k ✓\nAdd 2: [1,2,2,3,3] size=5>4 → remove 1. [2,2,3,3]\nAdd 4: [2,2,3,3,4]>4 → remove 2. [2,3,3,4]\nAdd 5: ... → [3,3,4,5]\nAdd 5: ... → [3,4,5,5]\nAdd 6: ... → [4,5,5,6]\nReturn heap.peek()=4 ✅`,
      keyInsight: `O(n log k) time, O(k) space. The reliable O(n log k) approach — slower than Quickselect on average but has no worst case. Great when k is small.`,
    },

    'Sort Descending': {
      intuition: `Sort the entire array in descending order. The answer is at index k-1. Clean and obvious — O(n log n) but very easy to implement.`,
      steps: [
        `Sort nums descending.`,
        `Return nums[k - 1].`,
      ],
      example: `nums=[3,2,3,1,2,4,5,5,6], k=4\nSorted desc: [6,5,5,4,3,3,2,2,1]\nnums[3]=4 ✅`,
      keyInsight: `O(n log n) time, O(1) space. Easiest to write. Good first attempt; follow up with Quickselect or heap if the interviewer asks for an improvement.`,
    },

    'Max Heap (Extract k Times)': {
      intuition: `Build a max-heap from all elements. Extract the maximum k times. The kth extracted value is the kth largest.`,
      steps: [
        `Heapify nums into a max-heap.`,
        `Call extractMax() k times.`,
        `Return the last extracted value.`,
      ],
      example: `nums=[3,2,3,1,2,4,5,5,6], k=4\nMax-heap: [6,5,5,4,3,3,2,2,1]\nExtract: 6, 5, 5, 4 ← 4th extraction\nReturn 4 ✅`,
      keyInsight: `O(n + k log n) time, O(n) space. Building a max-heap takes O(n). Each extraction is O(log n). Worse than Min Heap of size k when k is small relative to n.`,
    },

    'nth_element via partial sort': {
      intuition: `Use the language's built-in partial sort or nth_element. In C++: nth_element(nums.begin(), nums.begin()+k-1, nums.end(), greater<int>()). In Python: heapq.nlargest(k, nums)[-1]. These use Quickselect internally.`,
      steps: [
        `C++: nth_element(begin, begin + k - 1, end, greater<int>()); return nums[k-1].`,
        `Python: return heapq.nlargest(k, nums)[-1].`,
      ],
      example: `nums=[3,2,3,1,2,4,5,5,6], k=4\nheapq.nlargest(4, nums) → [6,5,5,4]\n[-1] → 4 ✅`,
      keyInsight: `O(n) average via Quickselect internally. The most concise solution. Useful in a real coding environment; in an interview, implement Quickselect manually to demonstrate understanding.`,
    },
  },
}
